{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from Bio import SeqIO\n",
    "def parseFasta(data):\n",
    "    d = {fasta.id : str(fasta.seq) for fasta in SeqIO.parse(data, \"fasta\")}\n",
    "    pd.DataFrame([d])\n",
    "\n",
    "    s = pd.Series(d, name='Sequence')\n",
    "    s.index.name = 'ID'\n",
    "    s.reset_index()\n",
    "    return pd.DataFrame(s)\n",
    "\n",
    "from sklearn.feature_extraction.text import TfidfVectorizer\n",
    "def kmerXTable(s, a, b):\n",
    "    tfid_vector = TfidfVectorizer(analyzer='char', ngram_range=(a,b))\n",
    "    s_hat = tfid_vector.fit_transform(s.Sequence)\n",
    "    kmerNames = tfid_vector.get_feature_names()\n",
    "    kmers = s_hat.toarray()\n",
    "    return pd.DataFrame(kmers,columns=kmerNames, index = s.index)\n",
    "\n",
    "# not used\n",
    "from sklearn.cluster import KMeans\n",
    "def doKmeans(rState, nClusters, data):\n",
    "    km = KMeans(random_state = rState, n_clusters = nClusters)\n",
    "    km.fit(data)\n",
    "    return km\n",
    "\n",
    "# not used\n",
    "from sklearn.decomposition import PCA\n",
    "def doPCA(data):\n",
    "    embedding = PCA()\n",
    "    embedding.fit(data)\n",
    "    return embedding\n",
    "\n",
    "# not used\n",
    "import matplotlib.pyplot as plt\n",
    "def graphPCA(pca, data, labels):\n",
    "    show = pd.DataFrame(pca.transform(data))\n",
    "    ax = show.plot.scatter(x=0, y=1, style=\"o\", c=labels, cmap = \"viridis\", s=2)\n",
    "    plt.title('PCA')\n",
    "    return ax\n",
    "\n",
    "# not used\n",
    "from sklearn.manifold import TSNE\n",
    "def doTSNE(rState, nComponents, data):\n",
    "    tSNEembedding = TSNE(n_components= nComponents, random_state = rState)\n",
    "    tSNEembedding_low = tSNEembedding.fit_transform(data)\n",
    "    return tSNEembedding_low\n",
    "\n",
    "# not used\n",
    "def graphTSNE(tSNEembedding_low, labels):\n",
    "    tSNEshow = pd.DataFrame(tSNEembedding_low, labels)\n",
    "    ax = tSNEshow.plot.scatter(x=0, y=1, style=\"o\", c=labels, cmap = \"viridis\", s=2)\n",
    "    plt.title('tSNE')\n",
    "    return ax\n",
    "\n",
    "# other imports \n",
    "import matplotlib.patches as mpatches\n",
    "from sklearn.cluster import MeanShift\n",
    "import warnings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "path1 = \"data/Sclerotinia_biocontrol_mycovirus_nucleotide.fasta\"\n",
    "path01 = \"data/mycovirus_genbank_all_refseq_nucleotide_unique.fasta\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "def kmeansPCA(fastaList):\n",
    "    # read in fasta files\n",
    "    virus1 = parseFasta(path1)\n",
    "    virus01 = parseFasta(path01)\n",
    "    virus01 = virus01.append(virus1)\n",
    "    virus01 = virus01.drop_duplicates(keep=\"last\")\n",
    "\n",
    "    # make kmer tables\n",
    "    kmer7Table1 = kmerXTable(virus1, 7,7)\n",
    "    kmer7Table01 = kmerXTable(virus01, 7,7)\n",
    "\n",
    "    # only use columns that confirmed virus killers have no zeros in for kmer length 7\n",
    "    cols = kmer7Table1.loc[:, (kmer7Table1 == 0).any(axis=0) != True].columns\n",
    "\n",
    "    #Kmeans\n",
    "    km = KMeans(random_state = 42, n_clusters = 2)\n",
    "    km.fit(kmer7Table01[cols])\n",
    "    \n",
    "    output = []\n",
    "    for i in range(0, len(fastaList)):\n",
    "        inputData = parseFasta(fastaList[i])\n",
    "        inputData[\"Sequence\"] = inputData[\"Sequence\"].apply(lambda x: x.replace(\"-\", \"\"))\n",
    "        kmer7TableInput = kmerXTable(inputData, 7,7)\n",
    "        \n",
    "        # generate array of labels\n",
    "        y_hat = km.predict(kmer7TableInput[cols])\n",
    "\n",
    "        #PCA\n",
    "        if len(inputData) > 1:\n",
    "            embedding = PCA()\n",
    "            embedding.fit(kmer7TableInput[cols])\n",
    "            show = pd.DataFrame(embedding.transform(kmer7TableInput[cols]))\n",
    "            # show kmeans clustering\n",
    "            show[\"labels\"] = y_hat\n",
    "            ax = show[show[\"labels\"]==1].plot.scatter(x=0, y=1, style=\"o\", color=\"red\", s=2)\n",
    "            show[show[\"labels\"]==0].plot.scatter(x=0, y=1, style=\"o\", color=\"blue\", s=2, ax=ax)\n",
    "            red = mpatches.Patch(color='red', label='Fungus Killers')\n",
    "            blue = mpatches.Patch(color='blue', label='Fungus Non-Killers')\n",
    "            plt.legend(handles=[red, blue])\n",
    "            plt.title('PCA Visualization for KMeans Clustering\\n' + fastaList[i])\n",
    "            plt.xlabel('First Principal Component')\n",
    "            plt.ylabel('Second Principal Component')\n",
    "            plt.savefig('nonNotebookFiles/kmeansPCA' + str(i) + '.png', bbox_inches='tight')\n",
    "            plt.close()\n",
    "\n",
    "        inputData[\"Labels\"] = y_hat\n",
    "        output.append(inputData)\n",
    "    return output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                                      Sequence  Labels\n",
      "ID                                                                    \n",
      "NC_000960.1  GGGGGTTAGAGAAATCTTGGGAGATTTCTATCGTCATAGACATATG...       1\n",
      "NC_001278.1  GGGAAATTTGTGAGATTATCGCCCTAAAGGATAACTCTCATGCGGG...       0\n",
      "NC_001492.1  GCCTATGGGTGGTCTACATAGGTGAGCATGCGTTGCTCGATATAGA...       1\n",
      "NC_001633.1  ACAAAATAATTGAAGAAATTATTTTTGTTTTACATATTTAGATCTT...       1\n",
      "NC_001641.1  GAATTTTTCGGTGAACCGGAATTATGTCGTCTCTGTTAAATTCATT...       0\n",
      "                                                      Sequence  Labels\n",
      "ID                                                                    \n",
      "NC_007415.1  CTTTAGTAAATACCACCGCCTCTTAATAAGTCTCGAGGCACTCCCC...       1\n",
      "NC_015939.1  GGGGTGATGGGAAACTTACGTTTCGCCAACGTTAAGGTACTGTGAT...       1\n",
      "NC_022896.1  TTTTTGGGGATGGTACTCTCAGGTTTGATCTTTGTAGATCCTAACT...       1\n",
      "NC_026510.1  TTGGCTCCTGGAGACCGTTAGGTCCCCAGAAGCGGTTCAATAGAAC...       1\n",
      "NC_027138.1  GCAATAAAAAGCACAGCCGGAAGGCTTTCTTTTTATTGCCCAGTTT...       1\n"
     ]
    }
   ],
   "source": [
    "x = kmeansPCA([\"data/mycovirus_genbank_all_refseq_nucleotide_unique.fasta\",  \"data/Sclerotinia_biocontrol_mycovirus_nucleotide.fasta\"])\n",
    "for df in x:\n",
    "    print(df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "def kmeansTSNE(fastaList):\n",
    "    # read in fasta files\n",
    "    virus1 = parseFasta(path1)\n",
    "    virus01 = parseFasta(path01)\n",
    "    virus01 = virus01.append(virus1)\n",
    "    virus01 = virus01.drop_duplicates(keep=\"last\")\n",
    "\n",
    "    # make kmer tables\n",
    "    kmer7Table1 = kmerXTable(virus1, 7,7)\n",
    "    kmer7Table01 = kmerXTable(virus01, 7,7)\n",
    "\n",
    "    # only use columns that confirmed virus killers have no zeros in for kmer length 7\n",
    "    cols = kmer7Table1.loc[:, (kmer7Table1 == 0).any(axis=0) != True].columns\n",
    "\n",
    "    #Kmeans\n",
    "    km = KMeans(random_state = 42, n_clusters = 2)\n",
    "    km.fit(kmer7Table01[cols])\n",
    "    \n",
    "    output = []\n",
    "    for i in range(0, len(fastaList)):\n",
    "        inputData = parseFasta(fastaList[i])\n",
    "        inputData[\"Sequence\"] = inputData[\"Sequence\"].apply(lambda x: x.replace(\"-\", \"\"))\n",
    "        kmer7TableInput = kmerXTable(inputData, 7,7)\n",
    "        \n",
    "        # generate array of labels\n",
    "        y_hat = km.predict(kmer7TableInput[cols])\n",
    "\n",
    "        #tSNE\n",
    "        if len(inputData) > 1:\n",
    "            tSNEembedding = TSNE(n_components= 2, random_state = 0)\n",
    "            tSNEembedding_low = tSNEembedding.fit_transform(kmer7TableInput[cols])\n",
    "            show = pd.DataFrame(tSNEembedding_low)\n",
    "            # show kmeans clustering\n",
    "            show[\"labels\"] = y_hat\n",
    "            ax = show[show[\"labels\"]==1].plot.scatter(x=0, y=1, style=\"o\", color=\"red\", s=2)\n",
    "            show[show[\"labels\"]==0].plot.scatter(x=0, y=1, style=\"o\", color=\"blue\", s=2, ax=ax)\n",
    "            red = mpatches.Patch(color='red', label='Fungus Killers')\n",
    "            blue = mpatches.Patch(color='blue', label='Fungus Non-Killers')\n",
    "            plt.legend(handles=[red, blue])\n",
    "            plt.title('tSNE Visualization for KMeans Clustering\\n' + fastaList[i])\n",
    "            plt.xlabel('First Component')\n",
    "            plt.ylabel('Second Component')\n",
    "            plt.savefig('nonNotebookFiles/kmeansTSNE' + str(i) + '.png', bbox_inches='tight')\n",
    "            plt.close()\n",
    "\n",
    "        inputData[\"Labels\"] = y_hat\n",
    "        output.append(inputData)\n",
    "    return output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                                      Sequence  Labels\n",
      "ID                                                                    \n",
      "NC_000960.1  GGGGGTTAGAGAAATCTTGGGAGATTTCTATCGTCATAGACATATG...       1\n",
      "NC_001278.1  GGGAAATTTGTGAGATTATCGCCCTAAAGGATAACTCTCATGCGGG...       0\n",
      "NC_001492.1  GCCTATGGGTGGTCTACATAGGTGAGCATGCGTTGCTCGATATAGA...       1\n",
      "NC_001633.1  ACAAAATAATTGAAGAAATTATTTTTGTTTTACATATTTAGATCTT...       1\n",
      "NC_001641.1  GAATTTTTCGGTGAACCGGAATTATGTCGTCTCTGTTAAATTCATT...       0\n",
      "                                                      Sequence  Labels\n",
      "ID                                                                    \n",
      "NC_007415.1  CTTTAGTAAATACCACCGCCTCTTAATAAGTCTCGAGGCACTCCCC...       1\n",
      "NC_015939.1  GGGGTGATGGGAAACTTACGTTTCGCCAACGTTAAGGTACTGTGAT...       1\n",
      "NC_022896.1  TTTTTGGGGATGGTACTCTCAGGTTTGATCTTTGTAGATCCTAACT...       1\n",
      "NC_026510.1  TTGGCTCCTGGAGACCGTTAGGTCCCCAGAAGCGGTTCAATAGAAC...       1\n",
      "NC_027138.1  GCAATAAAAAGCACAGCCGGAAGGCTTTCTTTTTATTGCCCAGTTT...       1\n"
     ]
    }
   ],
   "source": [
    "x = kmeansTSNE([\"data/mycovirus_genbank_all_refseq_nucleotide_unique.fasta\",  \"data/Sclerotinia_biocontrol_mycovirus_nucleotide.fasta\"])\n",
    "for df in x:\n",
    "    print(df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def kmeansMeanshiftPCA(fastaList):\n",
    "    # read in fasta files\n",
    "    virus1 = parseFasta(path1)\n",
    "    virus01 = parseFasta(path01)\n",
    "    virus01 = virus01.append(virus1)\n",
    "    virus01 = virus01.drop_duplicates(keep=\"last\")\n",
    "\n",
    "    # make kmer tables\n",
    "    kmer7Table1 = kmerXTable(virus1, 7,7)\n",
    "    kmer7Table01 = kmerXTable(virus01, 7,7)\n",
    "\n",
    "    # only use columns that confirmed virus killers have no zeros in for kmer length 7\n",
    "    cols = kmer7Table1.loc[:, (kmer7Table1 == 0).any(axis=0) != True].columns\n",
    "    \n",
    "    #meanshift\n",
    "    ms = MeanShift()\n",
    "    ms.fit(kmer7Table01[cols])\n",
    "    cluster_centers = ms.cluster_centers_\n",
    "\n",
    "    #Kmeans\n",
    "    with warnings.catch_warnings():\n",
    "        warnings.simplefilter(\"ignore\")\n",
    "        km = KMeans(init = cluster_centers, n_clusters = len(cluster_centers))\n",
    "        km.fit(kmer7Table01[cols])\n",
    "    newLabelClusters = km.predict(kmer7Table01[cols])[343:]\n",
    "    \n",
    "    output = []\n",
    "    for i in range(0, len(fastaList)):\n",
    "        inputData = parseFasta(fastaList[i])\n",
    "        inputData[\"Sequence\"] = inputData[\"Sequence\"].apply(lambda x: x.replace(\"-\", \"\"))\n",
    "        kmer7TableInput = kmerXTable(inputData, 7,7)\n",
    "        \n",
    "        # generate array of labels\n",
    "        y_hat = km.predict(kmer7TableInput[cols])\n",
    "        newLabels = []\n",
    "        for label in y_hat:\n",
    "            if label in newLabelClusters:\n",
    "                newLabels.append(1)\n",
    "            else:\n",
    "                newLabels.append(0)  \n",
    "\n",
    "        #PCA\n",
    "        if len(inputData) > 1:\n",
    "            embedding = PCA()\n",
    "            embedding.fit(kmer7TableInput[cols])\n",
    "            show = pd.DataFrame(embedding.transform(kmer7TableInput[cols]))\n",
    "            # show kmeans clustering\n",
    "            show[\"labels\"] = newLabels\n",
    "            ax = show[show[\"labels\"]==1].plot.scatter(x=0, y=1, style=\"o\", color=\"red\", s=2)\n",
    "            show[show[\"labels\"]==0].plot.scatter(x=0, y=1, style=\"o\", color=\"blue\", s=2, ax=ax)\n",
    "            red = mpatches.Patch(color='red', label='Fungus Killers')\n",
    "            blue = mpatches.Patch(color='blue', label='Fungus Non-Killers')\n",
    "            plt.legend(handles=[red, blue])\n",
    "            plt.title('PCA Visualization for KMeans Clustering with Mean Shift\\n' + fastaList[i])\n",
    "            plt.xlabel('First Principal Component')\n",
    "            plt.ylabel('Second Principal Component')\n",
    "            plt.savefig('nonNotebookFiles/kmeansMeanshiftPCA' + str(i) + '.png', bbox_inches='tight')\n",
    "            plt.close()\n",
    "\n",
    "        inputData[\"Labels\"] = newLabels\n",
    "        print(newLabels)\n",
    "        output.append(inputData)\n",
    "    return output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 1, 1, 0, 0, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 1, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]\n",
      "[1, 0, 1, 1, 1, 1, 0]\n",
      "                                                      Sequence  Labels\n",
      "ID                                                                    \n",
      "NC_000960.1  GGGGGTTAGAGAAATCTTGGGAGATTTCTATCGTCATAGACATATG...       1\n",
      "NC_001278.1  GGGAAATTTGTGAGATTATCGCCCTAAAGGATAACTCTCATGCGGG...       0\n",
      "NC_001492.1  GCCTATGGGTGGTCTACATAGGTGAGCATGCGTTGCTCGATATAGA...       0\n",
      "NC_001633.1  ACAAAATAATTGAAGAAATTATTTTTGTTTTACATATTTAGATCTT...       0\n",
      "NC_001641.1  GAATTTTTCGGTGAACCGGAATTATGTCGTCTCTGTTAAATTCATT...       0\n",
      "                                                      Sequence  Labels\n",
      "ID                                                                    \n",
      "NC_007415.1  CTTTAGTAAATACCACCGCCTCTTAATAAGTCTCGAGGCACTCCCC...       1\n",
      "NC_015939.1  GGGGTGATGGGAAACTTACGTTTCGCCAACGTTAAGGTACTGTGAT...       0\n",
      "NC_022896.1  TTTTTGGGGATGGTACTCTCAGGTTTGATCTTTGTAGATCCTAACT...       1\n",
      "NC_026510.1  TTGGCTCCTGGAGACCGTTAGGTCCCCAGAAGCGGTTCAATAGAAC...       1\n",
      "NC_027138.1  GCAATAAAAAGCACAGCCGGAAGGCTTTCTTTTTATTGCCCAGTTT...       1\n"
     ]
    }
   ],
   "source": [
    "x = kmeansMeanshiftPCA([\"data/mycovirus_genbank_all_refseq_nucleotide_unique.fasta\",  \"data/Sclerotinia_biocontrol_mycovirus_nucleotide.fasta\"])\n",
    "for df in x:\n",
    "    print(df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "def kmeansMeanshiftTSNE(fastaList):\n",
    "    # read in fasta files\n",
    "    virus1 = parseFasta(path1)\n",
    "    virus01 = parseFasta(path01)\n",
    "    virus01 = virus01.append(virus1)\n",
    "    virus01 = virus01.drop_duplicates(keep=\"last\")\n",
    "\n",
    "    # make kmer tables\n",
    "    kmer7Table1 = kmerXTable(virus1, 7,7)\n",
    "    kmer7Table01 = kmerXTable(virus01, 7,7)\n",
    "\n",
    "    # only use columns that confirmed virus killers have no zeros in for kmer length 7\n",
    "    cols = kmer7Table1.loc[:, (kmer7Table1 == 0).any(axis=0) != True].columns\n",
    "    \n",
    "    #meanshift\n",
    "    ms = MeanShift()\n",
    "    ms.fit(kmer7Table01[cols])\n",
    "    cluster_centers = ms.cluster_centers_\n",
    "\n",
    "    #Kmeans\n",
    "    with warnings.catch_warnings():\n",
    "        warnings.simplefilter(\"ignore\")\n",
    "        km = KMeans(init = cluster_centers, n_clusters = len(cluster_centers))\n",
    "        km.fit(kmer7Table01[cols])\n",
    "    newLabelClusters = km.predict(kmer7Table01[cols])[343:]\n",
    "    \n",
    "    output = []\n",
    "    for i in range(0, len(fastaList)):\n",
    "        inputData = parseFasta(fastaList[i])\n",
    "        inputData[\"Sequence\"] = inputData[\"Sequence\"].apply(lambda x: x.replace(\"-\", \"\"))\n",
    "        kmer7TableInput = kmerXTable(inputData, 7,7)\n",
    "        \n",
    "        # generate array of labels\n",
    "        y_hat = km.predict(kmer7TableInput[cols])\n",
    "        newLabels = []\n",
    "        for label in y_hat:\n",
    "            if label in newLabelClusters:\n",
    "                newLabels.append(1)\n",
    "            else:\n",
    "                newLabels.append(0)  \n",
    "\n",
    "        #PCA\n",
    "        if len(inputData) > 1:\n",
    "            tSNEembedding = TSNE(n_components= 2, random_state = 0)\n",
    "            tSNEembedding_low = tSNEembedding.fit_transform(kmer7TableInput[cols])\n",
    "            show = pd.DataFrame(tSNEembedding_low)\n",
    "            # show kmeans clustering\n",
    "            show[\"labels\"] = newLabels\n",
    "            ax = show[show[\"labels\"]==1].plot.scatter(x=0, y=1, style=\"o\", color=\"red\", s=2)\n",
    "            show[show[\"labels\"]==0].plot.scatter(x=0, y=1, style=\"o\", color=\"blue\", s=2, ax=ax)\n",
    "            red = mpatches.Patch(color='red', label='Fungus Killers')\n",
    "            blue = mpatches.Patch(color='blue', label='Fungus Non-Killers')\n",
    "            plt.legend(handles=[red, blue])\n",
    "            plt.title('tSNE Visualization for KMeans Clustering with Mean Shift\\n' + fastaList[i])\n",
    "            plt.xlabel('First Component')\n",
    "            plt.ylabel('Second Component')\n",
    "            plt.savefig('nonNotebookFiles/kmeansMeanshiftTSNE' + str(i) + '.png', bbox_inches='tight')\n",
    "            plt.close()\n",
    "\n",
    "        inputData[\"Labels\"] = newLabels\n",
    "        output.append(inputData)\n",
    "    return output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                                      Sequence  Labels\n",
      "ID                                                                    \n",
      "NC_000960.1  GGGGGTTAGAGAAATCTTGGGAGATTTCTATCGTCATAGACATATG...       1\n",
      "NC_001278.1  GGGAAATTTGTGAGATTATCGCCCTAAAGGATAACTCTCATGCGGG...       0\n",
      "NC_001492.1  GCCTATGGGTGGTCTACATAGGTGAGCATGCGTTGCTCGATATAGA...       0\n",
      "NC_001633.1  ACAAAATAATTGAAGAAATTATTTTTGTTTTACATATTTAGATCTT...       0\n",
      "NC_001641.1  GAATTTTTCGGTGAACCGGAATTATGTCGTCTCTGTTAAATTCATT...       0\n",
      "                                                      Sequence  Labels\n",
      "ID                                                                    \n",
      "NC_007415.1  CTTTAGTAAATACCACCGCCTCTTAATAAGTCTCGAGGCACTCCCC...       1\n",
      "NC_015939.1  GGGGTGATGGGAAACTTACGTTTCGCCAACGTTAAGGTACTGTGAT...       0\n",
      "NC_022896.1  TTTTTGGGGATGGTACTCTCAGGTTTGATCTTTGTAGATCCTAACT...       1\n",
      "NC_026510.1  TTGGCTCCTGGAGACCGTTAGGTCCCCAGAAGCGGTTCAATAGAAC...       1\n",
      "NC_027138.1  GCAATAAAAAGCACAGCCGGAAGGCTTTCTTTTTATTGCCCAGTTT...       1\n"
     ]
    }
   ],
   "source": [
    "x = kmeansMeanshiftTSNE([\"data/mycovirus_genbank_all_refseq_nucleotide_unique.fasta\",  \"data/Sclerotinia_biocontrol_mycovirus_nucleotide.fasta\"])\n",
    "for df in x:\n",
    "    print(df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                                      Sequence  Labels\n",
      "ID                                                                    \n",
      "NC_000960.1  GGGGGTTAGAGAAATCTTGGGAGATTTCTATCGTCATAGACATATG...       1\n",
      "NC_001278.1  GGGAAATTTGTGAGATTATCGCCCTAAAGGATAACTCTCATGCGGG...       0\n",
      "NC_001492.1  GCCTATGGGTGGTCTACATAGGTGAGCATGCGTTGCTCGATATAGA...       1\n",
      "NC_001633.1  ACAAAATAATTGAAGAAATTATTTTTGTTTTACATATTTAGATCTT...       1\n",
      "NC_001641.1  GAATTTTTCGGTGAACCGGAATTATGTCGTCTCTGTTAAATTCATT...       0\n",
      "                                                      Sequence  Labels\n",
      "ID                                                                    \n",
      "NC_007415.1  CTTTAGTAAATACCACCGCCTCTTAATAAGTCTCGAGGCACTCCCC...       1\n",
      "NC_015939.1  GGGGTGATGGGAAACTTACGTTTCGCCAACGTTAAGGTACTGTGAT...       1\n",
      "NC_022896.1  TTTTTGGGGATGGTACTCTCAGGTTTGATCTTTGTAGATCCTAACT...       1\n",
      "NC_026510.1  TTGGCTCCTGGAGACCGTTAGGTCCCCAGAAGCGGTTCAATAGAAC...       1\n",
      "NC_027138.1  GCAATAAAAAGCACAGCCGGAAGGCTTTCTTTTTATTGCCCAGTTT...       1\n",
      "                                                      Sequence  Labels\n",
      "ID                                                                    \n",
      "NC_000960.1  GGGGGTTAGAGAAATCTTGGGAGATTTCTATCGTCATAGACATATG...       1\n",
      "NC_001278.1  GGGAAATTTGTGAGATTATCGCCCTAAAGGATAACTCTCATGCGGG...       0\n",
      "NC_001492.1  GCCTATGGGTGGTCTACATAGGTGAGCATGCGTTGCTCGATATAGA...       1\n",
      "NC_001633.1  ACAAAATAATTGAAGAAATTATTTTTGTTTTACATATTTAGATCTT...       1\n",
      "NC_001641.1  GAATTTTTCGGTGAACCGGAATTATGTCGTCTCTGTTAAATTCATT...       0\n",
      "                                                      Sequence  Labels\n",
      "ID                                                                    \n",
      "NC_007415.1  CTTTAGTAAATACCACCGCCTCTTAATAAGTCTCGAGGCACTCCCC...       1\n",
      "NC_015939.1  GGGGTGATGGGAAACTTACGTTTCGCCAACGTTAAGGTACTGTGAT...       1\n",
      "NC_022896.1  TTTTTGGGGATGGTACTCTCAGGTTTGATCTTTGTAGATCCTAACT...       1\n",
      "NC_026510.1  TTGGCTCCTGGAGACCGTTAGGTCCCCAGAAGCGGTTCAATAGAAC...       1\n",
      "NC_027138.1  GCAATAAAAAGCACAGCCGGAAGGCTTTCTTTTTATTGCCCAGTTT...       1\n",
      "[21  9 21  2 21 18  9]\n",
      "                                                      Sequence  Labels\n",
      "ID                                                                    \n",
      "NC_000960.1  GGGGGTTAGAGAAATCTTGGGAGATTTCTATCGTCATAGACATATG...       1\n",
      "NC_001278.1  GGGAAATTTGTGAGATTATCGCCCTAAAGGATAACTCTCATGCGGG...       0\n",
      "NC_001492.1  GCCTATGGGTGGTCTACATAGGTGAGCATGCGTTGCTCGATATAGA...       0\n",
      "NC_001633.1  ACAAAATAATTGAAGAAATTATTTTTGTTTTACATATTTAGATCTT...       0\n",
      "NC_001641.1  GAATTTTTCGGTGAACCGGAATTATGTCGTCTCTGTTAAATTCATT...       0\n",
      "                                                      Sequence  Labels\n",
      "ID                                                                    \n",
      "NC_007415.1  CTTTAGTAAATACCACCGCCTCTTAATAAGTCTCGAGGCACTCCCC...       0\n",
      "NC_015939.1  GGGGTGATGGGAAACTTACGTTTCGCCAACGTTAAGGTACTGTGAT...       1\n",
      "NC_022896.1  TTTTTGGGGATGGTACTCTCAGGTTTGATCTTTGTAGATCCTAACT...       1\n",
      "NC_026510.1  TTGGCTCCTGGAGACCGTTAGGTCCCCAGAAGCGGTTCAATAGAAC...       1\n",
      "NC_027138.1  GCAATAAAAAGCACAGCCGGAAGGCTTTCTTTTTATTGCCCAGTTT...       1\n",
      "                                                      Sequence  Labels\n",
      "ID                                                                    \n",
      "NC_000960.1  GGGGGTTAGAGAAATCTTGGGAGATTTCTATCGTCATAGACATATG...       1\n",
      "NC_001278.1  GGGAAATTTGTGAGATTATCGCCCTAAAGGATAACTCTCATGCGGG...       0\n",
      "NC_001492.1  GCCTATGGGTGGTCTACATAGGTGAGCATGCGTTGCTCGATATAGA...       0\n",
      "NC_001633.1  ACAAAATAATTGAAGAAATTATTTTTGTTTTACATATTTAGATCTT...       0\n",
      "NC_001641.1  GAATTTTTCGGTGAACCGGAATTATGTCGTCTCTGTTAAATTCATT...       0\n",
      "                                                      Sequence  Labels\n",
      "ID                                                                    \n",
      "NC_007415.1  CTTTAGTAAATACCACCGCCTCTTAATAAGTCTCGAGGCACTCCCC...       0\n",
      "NC_015939.1  GGGGTGATGGGAAACTTACGTTTCGCCAACGTTAAGGTACTGTGAT...       1\n",
      "NC_022896.1  TTTTTGGGGATGGTACTCTCAGGTTTGATCTTTGTAGATCCTAACT...       1\n",
      "NC_026510.1  TTGGCTCCTGGAGACCGTTAGGTCCCCAGAAGCGGTTCAATAGAAC...       1\n",
      "NC_027138.1  GCAATAAAAAGCACAGCCGGAAGGCTTTCTTTTTATTGCCCAGTTT...       1\n"
     ]
    }
   ],
   "source": [
    "x = kmeansPCA([\"data/mycovirus_genbank_all_refseq_nucleotide_unique.fasta\",  \"data/Sclerotinia_biocontrol_mycovirus_nucleotide.fasta\"])\n",
    "for df in x:\n",
    "    print(df.head())\n",
    "    \n",
    "x = kmeansTSNE([\"data/mycovirus_genbank_all_refseq_nucleotide_unique.fasta\",  \"data/Sclerotinia_biocontrol_mycovirus_nucleotide.fasta\"])\n",
    "for df in x:\n",
    "    print(df.head())\n",
    "    \n",
    "x = kmeansMeanshiftPCA([\"data/mycovirus_genbank_all_refseq_nucleotide_unique.fasta\",  \"data/Sclerotinia_biocontrol_mycovirus_nucleotide.fasta\"])\n",
    "for df in x:\n",
    "    print(df.head())\n",
    "    \n",
    "x = kmeansMeanshiftTSNE([\"data/mycovirus_genbank_all_refseq_nucleotide_unique.fasta\",  \"data/Sclerotinia_biocontrol_mycovirus_nucleotide.fasta\"])\n",
    "for df in x:\n",
    "    print(df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
